core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/scheduling/WorkflowPipelinedRegions.scala

```diff
@@ -0,0 +1,166 @@
+      if (inputProcessingOrderForOp != null && inputProcessingOrderForOp.length > 1) {
+        for (i <- 1 to inputProcessingOrderForOp.length - 1) {
+          val prevInOrder = getPipelinedRegionFromOperatorId(
+            toOperatorIdentity(inputProcessingOrderForOp(i - 1).from)
+          )
+          val nextInOrder =
+            getPipelinedRegionFromOperatorId(toOperatorIdentity(inputProcessingOrderForOp(i).from))
+
+          if (
+            prevInOrder.getId() != nextInOrder.getId() && (!regionDependence.contains(
+              nextInOrder
+            ) || !regionDependence(nextInOrder).contains(prevInOrder))
+          ) {
+            val dependencies = regionDependence.getOrElseUpdate(
+              nextInOrder,
+              new ArrayBuffer[PipelinedRegion]()
+            )
+            dependencies.append(prevInOrder)
+            regionDependence(nextInOrder) = dependencies
+          }
+        }
+      }
+
+      // 2. Find dependencies due to blocking operators.
+      // e.g. The region before and after a sort operator has dependencies
+      val upstreamOps = workflow.getDirectUpstreamOperators(opId)
+      upstreamOps.foreach(upstreamOp => {
+        val linkFromUpstreamOp = LinkIdentity(
+          workflow.getOperator(upstreamOp).topology.layers.last.id,
+          workflow.getOperator(opId).topology.layers.head.id
+        )
+        if (workflow.getOperator(opId).isInputBlocking(linkFromUpstreamOp)) {
+          val prevInOrder = getPipelinedRegionFromOperatorId(upstreamOp)
+          val nextInOrder = getPipelinedRegionFromOperatorId(opId)
+          if (prevInOrder.getId() != nextInOrder.getId()) {
+            if (
+              !regionDependence.contains(
+                nextInOrder
+              ) || !regionDependence(nextInOrder).contains(prevInOrder)
+            ) {
+              val dependencies = regionDependence.getOrElseUpdate(
+                nextInOrder,
+                new ArrayBuffer[PipelinedRegion]()
+              )
+              dependencies.append(prevInOrder)
+              regionDependence(nextInOrder) = dependencies
+            }
+
+            if (
+              !regionTerminalOperatorInOtherRegions.contains(
+                prevInOrder
+              ) || !regionTerminalOperatorInOtherRegions(prevInOrder).contains(opId)
+            ) {
+
+              val terminalOps = regionTerminalOperatorInOtherRegions.getOrElseUpdate(
+                prevInOrder,
+                new ArrayBuffer[OperatorIdentity]()
+              )
+              terminalOps.append(opId)
+              regionTerminalOperatorInOtherRegions(prevInOrder) = terminalOps
+            }
+
+          }
+        }
+      })
+    })
+
+    for ((region, dependence) <- regionDependence) {
+      dependence.foreach(upstreamRegion =>
+        pipelinedRegionsDAG.addEdge(
+          upstreamRegion,
+          region,
+          PipelinedRegionDAGLink(upstreamRegion.getId(), region.getId())
```

Comment:
we can directly add the edges to the dag, instead of adding it to an array first

Reply:
done